llvm/unittests/ADT/STLExtrasTest.cpp

```diff
@@ -532,6 +534,44 @@ TEST(STLExtrasTest, ConcatRangeADL) {
   EXPECT_THAT(concat<const int>(S0, S1), ElementsAre(1, 2, 3, 4));
 }
 
+TEST(STLExtrasTest, ConcatRangePtrToSameClass) {
+  some_namespace::some_struct S0{};
+  some_namespace::some_struct S1{};
+  SmallVector<some_namespace::some_struct *> V0{&S0};
+  SmallVector<some_namespace::some_struct *> V1{&S1, &S1};
+
+  // Dereferencing all iterators yields `some_namespace::some_struct *&`; no
+  // conversion takes place, `reference_type` is `some_namespace::some_struct
+  // *&`.
+  auto C = concat<some_namespace::some_struct *>(V0, V1);
+  static_assert(
+      std::is_same_v<decltype(*C.begin()), some_namespace::some_struct *&>);
+  EXPECT_THAT(C, ElementsAre(&S0, &S1, &S1));
+  // `reference_type` should still allow container modification.
+  llvm::for_each(C, [&S0](auto &i) {
+    if (i == &S0)
+      i = nullptr;
+  });
```

Comment:
Could we use a regular range for loop?

Reply:
It was meant to be a one-liner, but `clang-format` broke lines this way.  Fixed to use range-`for` :+1:.